diff --git a/kernel/drivers/video/msm/mdss/mdss_spi_panel.c b/kernel/drivers/video/msm/mdss/mdss_spi_panel.c
index 41ddd31..bf586d2 100755
--- a/kernel/drivers/video/msm/mdss/mdss_spi_panel.c
+++ b/kernel/drivers/video/msm/mdss/mdss_spi_panel.c
@@ -408,6 +408,7 @@ int mdss_spi_panel_kickoff(struct mdss_panel_data *pdata,
 		scan_count++;
 	}
 
+/*	
 	INIT_COMPLETION(ctrl_pdata->spi_panel_te);
 
 	rc = wait_for_completion_timeout(&ctrl_pdata->spi_panel_te,
@@ -415,7 +416,7 @@ int mdss_spi_panel_kickoff(struct mdss_panel_data *pdata,
 
 	if (rc == 0)
 		pr_err("wait panel TE time out\n");
-
+*/
 	rc = mdss_spi_tx_pixel(tx_buf, ctrl_pdata->byte_pre_frame);
 
 	return rc;
@@ -1093,6 +1094,8 @@ static int mdss_spi_panel_regulator_init(struct platform_device *pdev)
 	return rc;
 
 }
+
+/*
 irqreturn_t spi_panel_te_handler(int irq, void *data)
 {
 	struct spi_panel_data *ctrl_pdata =
@@ -1105,7 +1108,7 @@ irqreturn_t spi_panel_te_handler(int irq, void *data)
 	complete(&ctrl_pdata->spi_panel_te);
 	return IRQ_HANDLED;
 }
-
+*/
 
 static struct device_node *mdss_spi_pref_prim_panel(
 		struct platform_device *pdev)
@@ -1150,13 +1153,14 @@ int spi_panel_device_register(struct device_node *pan_node,
 
 	pinfo->panel_max_fps = mdss_panel_get_framerate(pinfo);
 	pinfo->panel_max_vtotal = mdss_panel_get_vtotal(pinfo);
-
+/*
 	ctrl_pdata->disp_te_gpio = of_get_named_gpio(ctrl_pdev->dev.of_node,
 		"qcom,platform-te-gpio", 0);
 
 	if (!gpio_is_valid(ctrl_pdata->disp_te_gpio))
 		pr_err("%s:%d, TE gpio not specified\n",
 						__func__, __LINE__);
+*/
 
 	ctrl_pdata->rst_gpio = of_get_named_gpio(ctrl_pdev->dev.of_node,
 			 "qcom,platform-reset-gpio", 0);
@@ -1389,7 +1393,7 @@ static int mdss_spi_panel_probe(struct platform_device *pdev)
 
 	ctrl_pdata->panel_data.event_handler = mdss_spi_panel_event_handler;
 
-
+/*
 	init_completion(&ctrl_pdata->spi_panel_te);
 
 	rc = devm_request_irq(&pdev->dev,
@@ -1400,7 +1404,7 @@ static int mdss_spi_panel_probe(struct platform_device *pdev)
 		pr_err("TE request_irq failed.\n");
 		return rc;
 	}
-
+*/
 	pr_debug("%s: spi panel  initialized\n", __func__);
 	return 0;
 
diff --git a/kernel/drivers/video/msm/mdss/mdss_spi_panel.h b/kernel/drivers/video/msm/mdss/mdss_spi_panel.h
index 54e526a..f7a8246 100755
--- a/kernel/drivers/video/msm/mdss/mdss_spi_panel.h
+++ b/kernel/drivers/video/msm/mdss/mdss_spi_panel.h
@@ -83,11 +83,11 @@ struct spi_panel_data {
 	struct mdss_util_intf *mdss_util;
 	struct spi_pinctrl_res pin_res;
 	struct dss_module_power panel_power_data;
-	struct completion spi_panel_te;
+//	struct completion spi_panel_te;
 	int byte_pre_frame;
 	char *tx_buf;
 	u8 ctrl_state;
-	int disp_te_gpio;
+//	int disp_te_gpio;
 	int rst_gpio;
 	struct spi_panel_cmds on_cmds;
 	struct spi_panel_cmds off_cmds;
